### PR TITLE
OCD-3534 - Ability to create Surveillance Activity Report as Excel spreadsheet

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SurveillanceController.java [hunk 1]
```diff
@@ -494,6 +500,22 @@ public void download(@RequestParam(value = "type", required = false, defaultValu
...
+    public @ResponseBody ChplOneTimeTrigger getActivityReport(@RequestParam("start") String start, @RequestParam("end") String end) throws ValidationException, UserRetrievalException {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        LocalDate startDate;
+        LocalDate endDate;
+        try {
+            startDate = LocalDate.parse(start, formatter);
+            endDate = LocalDate.parse(end, formatter);
+            return survManager.submitActivityReportRequest(startDate, endDate);
+       } catch (DateTimeException e) {
+            throw new ValidationException("Could not parse the start and/or end date.  The correct format is 'yyyy-mm-dd'.");
```
Comment (kekey1) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SurveillanceController.java]: Move to errors.properties?

File: chpl/chpl-resources/src/main/resources-dev/quartz.properties [hunk 2]
```diff
@@ -1,4 +1,5 @@
 org.quartz.dataSource.openchpl.jndiURL=java:comp/env/jdbc/openchpl
+org.quartz.scheduler.jmx.export = true
```
Comment (kekey1) [inline chpl/chpl-resources/src/main/resources-dev/quartz.properties]: Can you just explain this change? Related to DataDog?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/ActivityReportActionPermission.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.stereotype.Component;
+
+import gov.healthit.chpl.permissions.domains.ActionPermissions;
+
+@Component("surveillanceActivityReportActionPermission")
+public class ActivityReportActionPermission extends ActionPermissions {
+
+    @Override
+    public boolean hasAccess() {
+       return getResourcePermissions().isUserRoleAdmin() || getResourcePermissions().isUserRoleOnc();
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/ActivityReportActionPermission.java]: This says ADMIN or ONC but the jobs.xml file says ADMIN or ACB

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/SurveillanceActivityReportDataGatherer.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.stereotype.Component;
+
+import gov.healthit.chpl.manager.SurveillanceManager;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2(topic = "surveillanceActivityReportJobLogger")
+@Component
+public class SurveillanceActivityReportDataGatherer {
+
+    @Autowired
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/SurveillanceActivityReportDataGatherer.java]: Use Autowired constructor instead

**Review by kekey1 — CHANGES_REQUESTED** [review]

File: chpl/chpl-resources/src/main/resources-dev/quartz.properties (hunk 2 above)
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources-dev/quartz.properties]: Yes - I'm going to move this to the Datadog ticket.

**Review by kekey1 — DISMISSED** [review]

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SurveillanceController.java [hunk 5]
```diff
@@ -53,15 +57,18 @@
...
 import gov.healthit.chpl.logging.Loggable;
 import gov.healthit.chpl.manager.ActivityManager;
 import gov.healthit.chpl.manager.PendingSurveillanceManager;
 import gov.healthit.chpl.manager.SurveillanceManager;
 import gov.healthit.chpl.manager.impl.SurveillanceAuthorityAccessDeniedException;
 import gov.healthit.chpl.permissions.ResourcePermissions;
+import gov.healthit.chpl.util.ErrorMessageUtil;
 import gov.healthit.chpl.util.FileUtils;
 import gov.healthit.chpl.validation.surveillance.reviewer.AuthorityReviewer;
+import gov.healthit.chpl.web.controller.results.BooleanResult;
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SurveillanceController.java]: Is checkstyle right about the unused import?

File: chpl/chpl-resources/src/main/resources/jobs.xml [hunk 6]
```diff
@@ -712,5 +712,21 @@
...
+            <name>surveillanceReportingActivityJob</name>
+            <group>chplBackgroundJobs</group>
+            <description>Build Excel spreadsheet based on surveillances open during a time range</description>
+            <job-class>gov.healthit.chpl.scheduler.job.surveillancereportingactivity.SurveillanceReportingActivityJob</job-class>
+            <durability>true</durability>
+            <recover>true</recover>
+            <job-data-map>
+                <entry>
+                    <key>authorities</key>
+                    <value>ROLE_ADMIN;ROLE_ONC</value>
```
Comment (andlar) [inline chpl/chpl-resources/src/main/resources/jobs.xml]: This should have `ROLE_ONC_STAFF`, right?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/Statistics.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import java.time.temporal.ChronoUnit;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class Statistics {
+
+    private Statistics() {}
+
+    public static Integer getMinimum(List<Integer> values) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/Statistics.java:13]: Do we need this? Or can something in the Collections package do this without rolling our own?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/charts/ClosedSurveillanceDurationByQuartileChart.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+            BoxAndWhiskerRenderer renderer = new BoxAndWhiskerRenderer();
+            renderer.setFillBox(false);
+            renderer.setMaximumBarWidth(25);
+            renderer.setMaxOutlierVisible(false);
+            renderer.setMinOutlierVisible(false);
+            renderer.setMeanVisible(false);
+
+            CategoryPlot plot = new CategoryPlot(getData(surveillances), xAxis, yAxis, renderer);
+            //plot.setOrientation(PlotOrientation.HORIZONTAL);
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/charts/ClosedSurveillanceDurationByQuartileChart.java]: Still needed?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/Statistics.java (hunk 7 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/Statistics.java:13]: There is a possibility of there being null values in the list and this function handles them.

File: chpl/chpl-resources/src/main/resources/jobs.xml (hunk 6 above)
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources/jobs.xml]: I added it, but I don't think it does anything.  The purpose of the `authorities` in the job definition is so the front end "knows" which jobs a particular user can run/schedule.   For `chplBackGroundJobs` which are generally called by managers, the security resides in the manager.

**Review by kekey1 — DISMISSED** [review]

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/ActivityReportActionPermission.java (hunk 3 above)
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/ActivityReportActionPermission.java]: I thought we said, at the FR, that this job would be available for ROLE_ONC_STAFF, even if the UI doesn't give them a way to run it yet. This looks like it's not available

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/excel/CellHelper.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
+package gov.healthit.chpl.scheduler.job.surveillancereportingactivity.excel;
+
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.util.Date;
+
+import org.apache.poi.ss.usermodel.Cell;
+import org.apache.poi.ss.usermodel.CellStyle;
+
+public class CellHelper {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/surveillancereportingactivity/excel/CellHelper.java]: Do we care about these checkstyle issues? I think there was one here as well as one in one of the XML files

**Review by andlar — DISMISSED** [review]

**Review by kekey1 — DISMISSED** [review]

**Review by andlar — DISMISSED** [review]

**Review by kekey1 — DISMISSED** [review]

**Review by kekey1 — APPROVED** [review]

**Review by andlar — APPROVED** [review]

**Pull request merged**